File: backend/src/agents/providers/cohere.py
from __future__ import annotations

from typing import Any, AsyncGenerator, Optional

import httpx

from src.agents.providers.openai_compatible import OpenAICompatibleProvider
from src.agents.providers.base import ProviderStreamDelta
from src.schemas.providers import ProviderMetadata, ProviderModel


class CohereProvider(OpenAICompatibleProvider):
    async def list_models(self, api_key: str, base_url: Optional[str] = None) -> list[ProviderModel]:
        endpoint = f"{(base_url or self.metadata.default_base_url).rstrip('/')}/models"
        headers = self._headers(api_key)
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.get(endpoint, headers=headers)
            if response.status_code == 404:
                return self._fallback_models()
            response.raise_for_status()
            payload = response.json()

        items = payload.get("data", payload)
        if not isinstance(items, list):
            return self._fallback_models()

        models: list[ProviderModel] = []
        for item in items:
            model_id = item.get("id") or item.get("name")
            if not model_id:
                continue
            models.append(
                ProviderModel(
                    id=model_id,
                    provider=self.metadata.id,
                    label=model_id,
                    owned_by=item.get("owned_by") or item.get("provider") or item.get("architecture", {}).get("tokenizer"),
                    supports_tools=True,
                    context_window=item.get("context_length") or item.get("top_provider", {}).get("context_length") or item.get("max_context_window"),
                )
            )
        if not models:
            return self._fallback_models()
        models.sort(key=lambda model: model.label.lower())
        return models

    async def stream_chat_completion(
        self,
        *,
        api_key: str,
        model: str,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]],
        base_url: Optional[str] = None,
        temperature: float = 0.2,
    ) -> AsyncGenerator[ProviderStreamDelta, None]:
        endpoint = f"{(base_url or self.metadata.default_base_url).rstrip('/')}/chat/completions"
        headers = self._headers(api_key)
        payload: dict[str, Any] = {
            "model": model,
            "messages": messages,
            "tools": tools,
            "temperature": temperature,
            "stream": True,
        }
        if tools:
            payload["tool_choice"] = "auto"

        async with httpx.AsyncClient(timeout=httpx.Timeout(90.0, connect=30.0)) as client:
            async with client.stream("POST", endpoint, headers=headers, json=payload) as response:
                response.raise_for_status()
                async for event in self._iter_sse_events(response):
                    if event == "[DONE]":
                        break
                    choice = (event.get("choices") or [{}])[0]
                    delta = choice.get("delta") or {}
                    finish_reason = choice.get("finish_reason")
                    text = self._extract_text(delta.get("content"))
                    reasoning = self._extract_text(delta.get("reasoning") or delta.get("reasoning_content") or delta.get("reason") or "")
                    tool_calls = delta.get("tool_calls") or None
                    if text or reasoning or tool_calls or finish_reason:
                        yield ProviderStreamDelta(
                            text=text,
                            reasoning=reasoning,
                            tool_calls=tool_calls,
                            finish_reason=finish_reason,
                            raw=event,
                        )

    def _fallback_models(self) -> list[ProviderModel]:
        model_ids = [
            "command-a-plus-05-2026",
            "command-r7b-03-2026",
            "command-r-plus-05-2026",
            "command-r-05-2026",
            "command-a-powerful",
        ]
        return [
            ProviderModel(
                id=mid,
                provider=self.metadata.id,
                label=mid,
            )
            for mid in model_ids
        ]

File: curro-ai/src/agents/providers/cohere.ts
import { OpenAICompatibleProvider } from "./base.js";
import { applyReasoningEffort } from "./reasoning.js";
import type { ChatCompletionOptions, ProviderModel } from "./types.js";

/**
 * Cohere exposes an OpenAI-compatible surface at `/compatibility/v1`, but its
 * `/models` endpoint can 404 (or return an unexpected shape) and it rejects the
 * `parallel_tool_calls` field. We fall back to a curated model list and trim the body.
 */
class CohereProvider extends OpenAICompatibleProvider {
  override async listModels(apiKey: string, baseUrl?: string): Promise<ProviderModel[]> {
    let response: Response;
    try {
      response = await fetch(`${this.baseUrl(baseUrl)}/models`, {
        method: "GET",
        headers: this.headers(apiKey),
      });
    } catch {
      return this.fallbackModels();
    }

    if (response.status === 404 || !response.ok) {
      return this.fallbackModels();
    }

    let payload: unknown;
    try {
      payload = await response.json();
    } catch {
      return this.fallbackModels();
    }

    const items: Array<Record<string, unknown>> = Array.isArray(payload)
      ? (payload as Array<Record<string, unknown>>)
      : (((payload as Record<string, unknown>)?.data as Array<Record<string, unknown>>) ?? []);
    if (!Array.isArray(items) || items.length === 0) {
      return this.fallbackModels();
    }

    const models: ProviderModel[] = [];
    for (const item of items) {
      const id = (item.id as string) || (item.name as string);
      if (!id) continue;
      models.push({
        id,
        provider: this.metadata.id,
        label: id,
        owned_by: (item.owned_by as string) || (item.provider as string) || null,
        context_window:
          (item.context_length as number) || (item.max_context_window as number) || null,
      });
    }
    if (models.length === 0) return this.fallbackModels();
    models.sort((a, b) => a.label.toLowerCase().localeCompare(b.label.toLowerCase()));
    return models;
  }

  protected override buildRequestBody(options: ChatCompletionOptions): Record<string, unknown> {
    const hasTools = Array.isArray(options.tools) && options.tools.length > 0;
    const body: Record<string, unknown> = {
      model: options.model,
      messages: options.messages,
      ...(hasTools ? { tools: options.tools, tool_choice: "auto" } : {}),
      temperature: options.temperature ?? 0.2,
      stream: true,
    };
    return applyReasoningEffort(body, options.effort);
  }

  private fallbackModels(): ProviderModel[] {
    const ids = [
      "command-a-plus-05-2026",
      "command-r7b-03-2026",
      "command-r-plus-05-2026",
      "command-r-05-2026",
      "command-a-powerful",
    ];
    return ids.map((id) => ({
      id,
      provider: this.metadata.id,
      label: id,
      owned_by: "cohere",
      context_window: null,
    }));
  }
}

export const cohereProvider = new CohereProvider({
  id: "cohere",
  label: "Cohere",
  defaultBaseUrl: "https://api.cohere.ai/compatibility/v1",
});
